toolchain/check/testdata/function/declaration/extern.carbon

```diff
@@ -0,0 +1,118 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// AUTOUPDATE
+
+// --- basic.carbon
+
+library "basic" api;
+
+extern fn F();
+
+// --- fail_redecl.carbon
+
+library "redecl" api;
+
+extern fn F();
+// CHECK:STDERR: fail_redecl.carbon:[[@LINE+6]]:1: ERROR: Redundant redeclaration of function F.
+// CHECK:STDERR: extern fn F();
+// CHECK:STDERR: ^~~~~~~~~~~~~~
+// CHECK:STDERR: fail_redecl.carbon:[[@LINE-4]]:1: Previously declared here.
+// CHECK:STDERR: extern fn F();
+// CHECK:STDERR: ^~~~~~~~~~~~~~
+extern fn F();
+
+// --- fail_redecl_extern.carbon
+
+library "redecl_extern" api;
+
+extern fn F();
+// CHECK:STDERR: fail_redecl_extern.carbon:[[@LINE+6]]:1: ERROR: Redundant redeclaration of function F.
+// CHECK:STDERR: fn F();
+// CHECK:STDERR: ^~~~~~~
+// CHECK:STDERR: fail_redecl_extern.carbon:[[@LINE-4]]:1: Previously declared here.
+// CHECK:STDERR: extern fn F();
+// CHECK:STDERR: ^~~~~~~~~~~~~~
+fn F();
+
+// --- fail_member_extern.carbon
+
+class C {
+  // CHECK:STDERR: fail_member_extern.carbon:[[@LINE+3]]:3: ERROR: `extern` not allowed on `fn` declaration on member functions.
```

Comment:
The wording of the diagnostic with the context string seems a bit awkward here.

I feel like we really want a fairly customized description of the entity with the invalid modifier:

- `fn` declaration
- member `fn` declaration
- `fn` definition (see below)

Not sure whether its best to have dedicated APIs for each of these entities or if there is a good way to let the caller customize the string here given that I think `fn` is coming from the token itself.

That said, as it will expand changes in the modifiers, happy if you'd rather leave some TODOs and fix this in a future iteration.